### DISCUSSION TITLE
Possible to keep a component among different URLs?

### DISCUSSION THREAD

**Discussion by borispoehland** [vercel/next.js#34493] [Help]
Hello guys,

I have the following problem: 

Do you know that feature from websites when you can toggle the view? E.g. List view <-> Grid view, something like that.

I want to implement this, but I want the 2 views to have different URLs.

One for example `/items/list` and one `/items/coverflow`.

But they both share a component that I don't want to re-render when switching URLs.

I basically want to change URLs and only switching the view. Keep the nav, keep the footer and the shared components despite URL change.

Is this possible with React/Next.js? Or do I have to toggle the view differently, e.g. with query params?

**Comment by SimonAlmers** [comment]
Hi, haven't tested this thought yet, but I came to think of a solution you could try.

Let's assume you have one "View Component" containing all your components, both the shared ones as well as the components you wish to toggle between. This View component would take a parameter determining which layout should be used 
```
<ViewComponent initialLayout={"grid"} />
```

You would then have two page files to generate the routes you want:
- pages/items/list.tsx
- pages/items/coverflow.tsx

Each of these would return the "ViewComponent" with the appropriate "grid"/"list" as the `initialLayout` parameter value.

This will give you the basic function when routing to switch between the layouts.

Now for the good part! 
When you want to toggle between the layouts you can use shallowRouting to update the URL without actually triggering a route change, this would mean that the URL will change but the user will remain on the current page.
You can then use a local state variable to manage the actual layout when toggling between the layouts. 

The initial value of this state would be the `initialLayout` param value. This means that if the user toggles from the list to grid view the page will not re-render but the URL will update so if the user manually refreshes the page the initialLayout will be set by the correct route.

Example state:
```
const ViewComponent = ({initialLayout}:{initialLayout: "grid" | "list"}) => {
  const [layout, setLayout] = useState(initialLayout)

return ...
}
```

You can find more info about shallow routing here: https://nextjs.org/docs/routing/shallow-routing

Let me know if something wasn't understandable/confusing and I'll try my best to clarify.

Wrote this on my smartphone so might have missed something.

✌🏻😃
/Simon

**Reply by borispoehland** [reply]
Hey buddy,

Thanks for the answer first of all 🤩

I always thought shallow routing means that if you switch to another page, you don't trigger their data fetching methods.

But what you are suggesting is basically that this also means that you can change the URL without causing a re-render?

I will try this tomorrow :)

**Reply by SimonAlmers** [reply]
My guess is that it would work, haven't had time to test this right now but fairly certain that I've made this work earlier on a similar mission ✌🏻😃

I often use this technique for pages that have some sort of search/filter components where I use shallow routing to update the current URL with the search & filter state represented as query params. This means that on first load I hydrate the search and filter state from the query params and in this way users never have to "loose" state if they accidentally refresh the search page and many of our customers also like to bookmark/save the search URL for searches they frequently visit.

**Reply by borispoehland** [reply]
But wait!

I have 2 pages then:
* `/pages/items/list` and
* `/pages/items/coverflow`

I have a Layout component, but that only contains Nav and Footer. So at least they will not re-render as they are one layer higher (`_app.js`)

But now, both these pages have a shared component that is **not** part of the layout. 

E.g.
```
// pages/items/list

const List = () => {
   return (
      <SharedComponent/>
      <ViewComponent initialLayout="list"/>
   );
}
```
```
// pages/items/coverflow

const Coverflow = () => {
   return (
      <SharedComponent/>
      <ViewComponent initialLayout="coverflow"/>
   );
}
```

Then the `<SharedComponent/>` will be re-rendered, right? Because there is no way for React to detect that this is the same component that was used on the sibling route.

That was basically my original question. If there's a way to tell React/Next.js: "Hey, on this new Url ONLY the `<ViewComponent/>` changed, so only replace that in the DOM and not render completely new and thus overwrite the `<SharedComponent/>` with itself"
